3. Create a function named get_db_url. It should accept a 
username, hostname, password, and database name and return 
a url connection string formatted like in the example at the start of this lesson.

In [1]:
from env import host, user, password

In [11]:
import pandas as pd
import numpy as np

In [6]:
host = host
username = user
password = password

def get_db_url(db_name, host=host, user=username, password=password):
    url = f'mysql+pymysql://{user}:{password}@{host}/{db_name}'
    return url

4. Use your function to obtain a connection to the employees database.

In [15]:
connection = get_db_url('employees')

5. Once you have successfully run a query:

In [18]:
pd.read_sql('''
SELECT *
FROM employees
LIMIT 5;
''', connection)

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

a. Intentionally make a typo in the database url. What kind of error message do you see?

In [19]:
bad_connection = get_db_url('employ')

pd.read_sql('''
SELECT *
FROM employees
LIMIT 5;
''', bad_connection)

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'oneil_2093'@'%' to database 'employ'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

# What kind of error message do you see?
(pymysql.err.OperationalError) (
    1044, "Access denied for user 'oneil_2093'@'%' to database 'employ'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

b. Intentionally make an error in your SQL query. What does the error message look like?

In [20]:
pd.read_sql('''
SELECT *
FROM employ
LIMIT 5;
''', connection)

ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employ' doesn't exist")
[SQL: 
SELECT *
FROM employ
LIMIT 5;
]
(Background on this error at: https://sqlalche.me/e/14/f405)

# What does the error message look like?
ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employ' doesn't exist")
[SQL: 
SELECT *
FROM employ
LIMIT 5;
]
(Background on this error at: https://sqlalche.me/e/14/f405)

6. Read the employees and titles tables into two separate DataFrames.

In [25]:
employees_query = '''
SELECT *
FROM employees;
'''

employees_df = pd.read_sql(employees_query, connection)
employees_df.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [26]:
titles_query = '''
SELECT *
FROM titles;
'''

titles_df = pd.read_sql(titles_query, connection)
titles_df.head()

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

7. How many rows and columns do you have in each DataFrame? Is that what you expected?

In [27]:
employees_df.shape
# 300,024 rows, 6 columns
# expected 300,024 rows, 6 columns

(300024, 6)

In [28]:
titles_df.shape
# 443,308 rows, 4 columns
# expected 443,308 rows, 4 columns

(443308, 4)

8. Display the summary statistics for each DataFrame.

In [31]:
employees_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [32]:
titles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


9. How many unique titles are in the titles DataFrame?

In [36]:
len(titles_df.title.unique())

7

10. What is the oldest date in the to_date column?

In [37]:
min(titles_df.to_date)

datetime.date(1985, 3, 1)

11. What is the most recent date in the to_date column?

In [45]:
import datetime

In [53]:
max(titles_df.to_date[titles_df.to_date < datetime.date.today()])

datetime.date(2002, 8, 1)